### ISSUE TITLE
Document usage of server creation in typescript 

### ISSUE THREAD

**Issue by timReynolds** [apollographql/apollo-server#1606]
Apollo server doesn't provide any documentation for usage with TypeScript, whilst this wouldn't normally be an issue for most projects the server uses some advanced types which aren't obvious to consume. 

Example 
```
const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: (): **???** => {
    return {
      blogPostApi: new BlogPostApi(
        config.prismic.url,
        config.prismic.accessToken
      )
    };
  }
});
```

First pass this looks like it should be `DataSource<IDataSources>` where IDataSource is something like;

```
interface IDataSources {
  blogPostApi: IBlogPostApi;
}
```

But returns;
```
src/index.ts:47:7 - error TS2322: Type '{ blogPostApi: BlogPostApi; }' is not assignable to type 'DataSource<IDataSources>'.
  Object literal may only specify known properties, and 'blogPostApi' does not exist in type 'DataSource<IDataSources>'.
```

**Comment by nareshbhatia** [comment]
It is strange that every working template I have seen uses `ts-node` and not `tsc`. `ts-node` seems to have no issues running apollo-server, but the moment I try to do a production build using `tsc`, I start getting compiler errors. Would really help if someone can share a working example with `tsc`.

If it helps, I have a simple server that runs perfectly well with `ts-node` but not with `tsc`. Here's the repo: https://github.com/nareshbhatia/graphql-bookstore. I managed to get the compiler errors down to zero, but in the process had to set the TypeScript target to `es2016`. This obviously does not work with node because the emitted code contains `import` statements.

- To run the server with ts-node, just run `yarn dev`.
- To run in production mode, first do a build using `yarn build`. This should produce no errors. However, if you try to run using `node dist/index.js` you will immediately get an error from node because it cannot understand imports.

**Comment by JacksonKearl** [comment]
Hey @nareshbhatia, to use `tsc` with any Node project you need to set `"module": "commonjs"` in your `tsconfig.json`.

And @timReynolds I know this is pretty old, but `dataSources` actually receives a `() => DataSources<TContext>`, not `DataSource<TContext>`. The difference being `DataSources` is an index type mapping from `name`'s to `DataSource`'s.

**Comment by lookfirst** [comment]
@JacksonKearl This issue is still not finished. The problem I'm having is that DataSources isn't exported publicly so there is no typesafe way to define this stuff. 

I'm using Apollo though `apollo-server-cloud-functions`

**Comment by lookfirst** [comment]
Just added a comment in the above issue that documents things. Maybe someone can pull that as an example for the website.

**Comment by nareshbhatia** [comment]
@JacksonKearl, thanks for the suggestion to add `"module": "commonjs"` to tsconfig. This helps, but now I am blocked by another error:

```
/Users/narbhati/projects/graphql-bookstore/apollo-bookstore-server/node_modules/graphql/validation/validate.js:80
    throw new Error(errors.map(function (error) {
    ^

Error: Unknown type "Query".
```

However this seems to be related to another issue: https://github.com/apollographql/apollo-server/issues/2332

**Comment by nareshbhatia** [comment]
Fixed the issue mentioned above. It was easy! I was not copying the typedef files to the dist folder. This was required for [merge-graphql-schemas](https://github.com/Urigo/merge-graphql-schemas) to stitch the modules together.

If anyone is interested in the complete working solution, you can see my repo [here](https://github.com/nareshbhatia/graphql-bookstore/tree/master/apollo-bookstore-server).

**Comment by NickTomlin** [comment]
I'm struggling with this as well. The type information seems to indicate that a simple object of name -> DataSource instance is fine:

```
export declare type DataSources<TContext> = {
    [name: string]: DataSource<TContext>;
};
```

But the following:

```
server({ dataSources: () => ({ foo: new RestDataSource() })}
```

results in an error:

```
foo not assignable to type 'DataSources<object>`
```

**Comment by lookfirst** [comment]
```
import {DataSource} from "apollo-datasource";
export MyDataSource extends DataSource {}
```

```
// this is a fugly import imho...
import {DataSources} from "apollo-server-core/dist/graphqlOptions";

export interface IDataSources {
	myDS: MyDataSource;
}

export function buildDataSources() {
	return {
		myDS: new MyDataSource(),
	} as DataSources<IDataSources>;
}
```

```
new ApolloServer({
	dataSources: () => buildDataSources(),
})
```

Not shown here, but I pass the base datasource into buildDataSources() and then through to the MyDataSource wrapper object.

**Comment by glasser** [comment]
The Apollo Server 4 docs use TypeScript by default (and also let you see code examples as JavaScript).

I agree with you that using dataSources was strange from a typing perspective (for example, it meant that the type of the context value returned from your `context` function didn't match the type of the context value available to resolvers and plugins); the dataSources option has been removed from Apollo Server 4 primarily for this reason.
